packages/teleport-monitoring/src/synchronizers/GenericSynchronizer.ts

```diff
@@ -0,0 +1,105 @@
+import { merge } from 'lodash'
+
+import { BlockchainClient } from '../peripherals/blockchain'
+import { SynchronizerStatusRepository } from '../peripherals/db/SynchronizerStatusRepository'
+import { TxHandle } from '../peripherals/db/utils'
+import { delay } from '../utils'
+
+export interface SyncOptions {
+  tipSyncDelay: number // ms
+  saveDistanceFromTip: number
+}
+
+export const defaultSyncOptions: SyncOptions = {
+  tipSyncDelay: 5_000,
+  saveDistanceFromTip: 0,
+}
+
+export type SynchronizerState = 'stopped' | 'syncing' | 'synced'
+
+export abstract class GenericSynchronizer {
+  private readonly options: SyncOptions
+  public readonly syncName: string
+
+  constructor(
+    private readonly blockchain: BlockchainClient,
+    private readonly synchronizerStatusRepository: SynchronizerStatusRepository,
+    public readonly domainName: string,
+    public readonly startingBlock: number,
+    public readonly blocksPerBatch: number,
+    _options?: Partial<SyncOptions>,
+  ) {
+    this.syncName = this.constructor.name
+    this.options = merge({}, defaultSyncOptions, _options)
```

Comment:
Should we require that `options.tipSyncDelay > 1_000` ? Otherwise `syncOnce()` could end up doing more than a single sync.

Reply:
Ah ok, fine then :)